### PR TITLE
Add fields for recommendation status

### PR DESCRIPTION
**Pull request by qmhu**
Add more fields for recommendation status to support different recommender's result.

## Field Explanation

- recommendedValue: previous field for recommendation. It will compatible for resource/replicas recommendation, but not supported for other recommenders
- targetRef: the target object info for recommend result. more of the time it is the same to spec.targetRef, but for hpa recommender it target to the hpa object while spec.targetRef is the workload object
- recommendedData: the recommended yaml for this resource. MergePatch yaml format. you can use it to patch update.
- currentData: the current yaml for this resource. MergePatch yaml format. you can use it to rollback update if needed.
- action: Apply/Patch/Delete. Action guide you how to finish this recommendation.
- description: describe information about this recommendation result. 

## What's Json MergePatch

https://datatracker.ietf.org/doc/html/rfc7386

Kustomize use similar function to manage yaml resource:https://kubernetes.io/zh-cn/docs/tasks/manage-kubernetes-objects/kustomization/#%E5%9F%BA%E5%87%86-bases-%E4%B8%8E%E8%A6%86%E7%9B%96-overlays (json patch and StrategicMergePatch)

The recommendedData/currentData provide a yaml patch that user can use it to patch object or unmarshal it and get any fields what you want, let's show a sample code about how to use it:

```
deploys := []appsv1.Deployment{
{
	ObjectMeta: metav1.ObjectMeta{
		Name: "pod1",
	},
	Spec: appsv1.DeploymentSpec{
		Template: v1.PodTemplateSpec{
			Spec: v1.PodSpec{
				Containers: []v1.Container{{
					Name:  "container1",
					Image: "nginx",
					Resources: v1.ResourceRequirements{
						Requests: map[v1.ResourceName]resource.Quantity{
							v1.ResourceCPU:    *resource.NewQuantity(1, resource.DecimalSI),
							v1.ResourceMemory: *resource.NewQuantity(10, resource.DecimalSI),
						},
					},
				}, {
					Name:  "container2",
					Image: "nginx",
					Resources: v1.ResourceRequirements{
						Requests: map[v1.ResourceName]resource.Quantity{
							v1.ResourceCPU:    *resource.NewQuantity(1, resource.DecimalSI),
							v1.ResourceMemory: *resource.NewQuantity(10, resource.DecimalSI),
						},
					},
				}},
			},
		},
	},
},
{
	ObjectMeta: metav1.ObjectMeta{
		Name: "pod1",
	},
	Spec: appsv1.DeploymentSpec{
		Template: v1.PodTemplateSpec{
			Spec: v1.PodSpec{
				Containers: []v1.Container{{
					Name: "container1",
					Resources: v1.ResourceRequirements{
						Requests: map[v1.ResourceName]resource.Quantity{
							v1.ResourceCPU:    *resource.NewQuantity(2, resource.DecimalSI),
							v1.ResourceMemory: *resource.NewQuantity(20, resource.DecimalSI),
						},
					},
				}, {
					Name: "container2",
					Resources: v1.ResourceRequirements{
						Requests: map[v1.ResourceName]resource.Quantity{
							v1.ResourceCPU:    *resource.NewQuantity(2, resource.DecimalSI),
							v1.ResourceMemory: *resource.NewQuantity(20, resource.DecimalSI),
						},
					},
				}},
			},
		},
	},
},
}

old, _ := json.Marshal(deploys[0])
new, _ := json.Marshal(deploys[1])

patchBytes, _ := jsonpatch.CreateMergePatch(old, new)
fmt.Println(patchBytes)
/*
{
		"spec": {
		  "template": {
		    "spec": {
		      "containers": [
		        {
		          "image": "nginx",
		          "name": "container1",
		          "resources": {
		            "requests": {
		              "cpu": "2",
		              "memory": "20"
		            }
		          }
		        },
		        {
		          "image": "nginx",
		          "name": "container2",
		          "resources": {
		            "requests": {
		              "cpu": "2",
		              "memory": "20"
		            }
		          }
		        }
		      ]
		    }
		  }
		}
}
*/

patchBytes1, _ := jsonpatch.CreateMergePatch(new, old)
fmt.Println(patchBytes1)
/*
{
	"spec": {
	  "template": {
	    "spec": {
	      "containers": [
	        {
	          "image": "nginx",
	          "name": "container1",
	          "resources": {
	            "requests": {
	              "cpu": "1",
	              "memory": "10"
	            }
	          }
	        },
	        {
	          "image": "nginx",
	          "name": "container2",
	          "resources": {
	            "requests": {
	              "cpu": "1",
	              "memory": "10"
	            }
	          }
	        }
	      ]
	    }
	  }
	}
}	
*/

result := appsv1.Deployment{}
err := json.Unmarshal(patchBytes, &result)     // 可以反序列化成Deployment，获取任意想拿到的字段
if err != nil {
fmt.Println(err)
}
fmt.Println(result.Name)
```

## Examples

Resource recommender
```
recommendedValue: 
  containers:
      - containerName: nginx
        target:
          cpu: 100m
          memory: 100Mi
targetRef:
  name: nginx
  namespace: default
  kind: Deployment
  apiVersion: v1
recommendedData: |
  spec:
    template:
      spec:
        containers:
          - image: nginx
            name: container1
            resources:
              requests:
                cpu: '1'
                memory: '10'
          - image: nginx
            name: container2
            resources:
              requests:
                cpu: '1'
                memory: '10'
currentData: |
  spec:
    template:
      spec:
        containers:
          - image: nginx
            name: container1
            resources:
              requests:
                cpu: '2'
                memory: '20'
          - image: nginx
            name: container2
            resources:
              requests:
                cpu: '1'
                memory: '20'
action: Update
description: Saving cost for deployment nginx.
```

Replicas recommender
```
recommendedValue: 
  replicas: 3
targetRef:
  name: nginx
  namespace: default
  kind: Deployment
  apiVersion: v1
recommendedData: |
  spec:
    replicas: 3
currentData: |
  spec:
    replicas: 5
action: Update
description: Saving cost for deployment nginx.
```

HPA recommender
```
recommendedValue: 
  effectiveHPA:
      maxReplicas: 3
      metrics:
      - resource:
          name: cpu
          target:
            averageUtilization: 75
            type: Utilization
        type: Resource
      minReplicas: 3
targetRef:
  name: nginx
  namespace: default
  kind: HorizontalPodAutoscaler
  apiVersion: v1
recommendedData: |
  apiVersion: autoscaling.crane.io/v1alpha1
  kind: EffectiveHorizontalPodAutoscaler
  metadata:
    name: php-apache
  spec:
    scaleTargetRef:
      apiVersion: apps/v1
      kind: Deployment
      name: php-apache
    minReplicas: 1 
    maxReplicas: 10 
    scaleStrategy: Auto
    metrics: 
    - type: Resource
      resource:
        name: cpu
        target:
          type: Utilization
          averageUtilization: 50
    prediction: 
      predictionWindowSeconds: 3600
      predictionAlgorithm:
        algorithmType: dsp
        dsp:
          sampleInterval: "60s"
          historyLength: "3d"
currentData: ""
action: Recreate
description: Saving cost for deployment nginx.
```

Storage recommender
```
recommendedValue: ""
targetRef:
  name: nginx
  namespace: default
  kind: PersistenceVolume
  apiVersion: v1
recommendedData: ""
currentData: ""
action: Delete
description: Idle persistence volume, suggest to delete it.
```

About the field name for recommendedData/currentData, maybe we can find something better, like: recommendedInfo/recommendedResult/recommendedContent.

### CODE REVIEW

**Review by xieydd — APPROVED** [review]

File: analysis/v1alpha1/types.go [hunk 1]
```diff
@@ -89,6 +88,22 @@ type RecommendationStatus struct {
 	LastUpdateTime *metav1.Time `json:"lastUpdateTime,omitempty"`
 }
 
+// RecommendationContent contains results for one recommendation
+type RecommendationContent struct {
+	// +optional
```
Comment (xieydd) [inline analysis/v1alpha1/types.go:93]: Can we add more description in here.

**Review by xieydd — CHANGES_REQUESTED** [review]

File: analysis/v1alpha1/types.go (hunk 1 above)
Comment (qmhu) [reply analysis/v1alpha1/types.go:93]: Description about this struct?

**Review by mfanjie — APPROVED** [review]
LGTM

File: analysis/v1alpha1/types.go (hunk 1 above)
Comment (xieydd) [reply analysis/v1alpha1/types.go:93]: Yep

**Pull request merged**